### PR TITLE
Support formatting LSP requests

### PR DESCRIPTION
**Pull request by msujew**
Adds an Xtext-like API to format documents. Does not support automated tests yet, as that would require some sort of virtual formatted document.

### CODE REVIEW

File: packages/langium/src/lsp/formatter.ts [hunk 1]
```diff
@@ -0,0 +1,633 @@
...
+    formatDocument(document: LangiumDocument, params: DocumentFormattingParams, cancelToken?: CancellationToken): MaybePromise<TextEdit[]>
+    formatDocumentRange(document: LangiumDocument, params: DocumentRangeFormattingParams, cancelToken?: CancellationToken): MaybePromise<TextEdit[]>
+    formatDocumentOnType(document: LangiumDocument, params: DocumentOnTypeFormattingParams, cancelToken?: CancellationToken): MaybePromise<TextEdit[]>
+}
+
+export abstract class AbstractFormattingService implements FormattingService {
+
+    protected collector: FormattingCollector = () => { /* Does nothing at first */ }
+
+    protected formatter<T extends AstNode>(node: T): Formatter<T> {
```
Comment (spoenemann) [inline packages/langium/src/lsp/formatter.ts:28]: Please document the methods that are meant to be used by subclasses.

File: packages/langium/src/validation/validation-registry.ts [hunk 2]
```diff
@@ -36,6 +36,8 @@ export type ValidationAcceptor = <N extends AstNode>(severity: 'error' | 'warnin
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export type ValidationCheck = (node: any, accept: ValidationAcceptor, cancelToken: CancellationToken) => MaybePromise<void>;
 
+export type ValidationChecks<T extends string> = { [type in T]?: ValidationCheck | ValidationCheck[] }
```
Comment (spoenemann) [inline packages/langium/src/validation/validation-registry.ts:39]: Please update all example languages as well as the Yeoman template to use this type.

**Review by spoenemann — COMMENTED** [review]
Great work, I like the new API!

File: examples/domainmodel/src/language-server/domain-model-formatting.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import * as ast from './generated/ast';
+
+export class DomainModelFormattingService extends AbstractFormattingService {
+
+    protected format(node: AstNode): void {
+        if (ast.isPackageDeclaration(node)) {
+            const formatter = this.formatter(node);
+            const bracesOpen = formatter.keyword('{');
+            const bracesClose = formatter.keyword('}');
+            formatter.interior(bracesOpen, bracesClose).prepend(Formatting.indent());
```
[the remark this answers is not in the document]
Comment (msujew) [reply examples/domainmodel/src/language-server/domain-model-formatting.ts:17]: Yes, exactly. `interior` selects every CST node between the `{` and `}` keyword. We then prepend each CST node an `indent` formatting, which automatically applies a line break if it does not exist yet.

File: packages/langium/src/lsp/language-server.ts [hunk 4]
```diff
@@ -186,6 +193,21 @@ export function addFoldingRangeHandler(connection: Connection, services: Langium
...
+        services
+    ));
+    connection.onDocumentRangeFormatting(createRequestHandler(
+        (services, document, params, cancelToken) => services.lsp.FormattingService?.formatDocumentRange(document, params, cancelToken),
+        services
+    ));
+    // connection.onDocumentOnTypeFormatting(createRequestHandler(
+    //     (services, document, params, cancelToken) => services.lsp.FormattingService?.formatDocumentOnType(document, params, cancelToken),
+    //     services
+    // ));
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/lsp/language-server.ts]: I implemented this one properly now.

File: examples/domainmodel/src/language-server/domain-model-formatting.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (spoenemann) [reply examples/domainmodel/src/language-server/domain-model-formatting.ts]: I don't get that yet. How do you decide where to apply a line break? We can't just do that for every leaf node.

**Review by spoenemann — APPROVED** [review]
:+1:

File: examples/domainmodel/src/language-server/domain-model-formatting.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (msujew) [reply examples/domainmodel/src/language-server/domain-model-formatting.ts]: It doesn't select **every** node, only the **direct descendants** between those two nodes. See the following example:

```ts
class x { // starting the `interior` call
  x = 42; // This expressions is the first selected node by the `interior` call
  /*
   * This is a JSDoc comment
   */  // This whole comment is the second selected node
  method() { ... } // This method is the third selected node
} // ending the `interior` call
```

In total this `interior` call only gathers 3 cst nodes, to which the line break and indentation is preprended. Does that make sense?

[the remark this answers is not in the document]
Comment (spoenemann) [reply examples/domainmodel/src/language-server/domain-model-formatting.ts]: Ok, I'm wondering whether this is sufficient to cover all cases. We can start with that and change it later if necessary.

File: packages/langium/src/lsp/formatter.ts [hunk 5]
```diff
@@ -0,0 +1,661 @@
...
+    allowLess?: boolean
+}
+
+export interface FormattingMove {
+    characters?: number
+    tabs?: number
+    lines?: number
+}
+
+export namespace Formatting {
```
[the remark this answers is not in the document]
Comment (msujew) [reply packages/langium/src/lsp/formatter.ts:672]: I added documentation to all exported classes/interfaces and their methods/functions

**Pull request merged**